[LBSE] Fix svg/repaint/image-with-clip-path.svg

rwlbuis opened the pull request:
#### 6dc9c6a25c09a9b226ac1d692b4dcad6b29be15b
<pre>
[LBSE] Fix svg/repaint/image-with-clip-path.svg
<a href="https://bugs.webkit.org/show_bug.cgi?id=270187">https://bugs.webkit.org/show_bug.cgi?id=270187</a>

Reviewed by Nikolas Zimmermann.

On a transform change from identity matrix to non-identity matrix
the render layer needs to become a stacking context without relying on
a style update in the case of LBSE. Add logic for this in repaintOrRelayoutAfterSVGTransformChange.

Also add a correction to only call updateLayerTransform once in
repaintOrRelayoutAfterSVGTransformChange.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateTransform):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::repaintOrRelayoutAfterSVGTransformChange):

Canonical link: <a href="https://commits.webkit.org/275683@main">https://commits.webkit.org/275683@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5511c47bc94b3ae4fd74e06cce2e06c0802a37cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42348 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21366 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44947 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38465 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44655 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24589 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18710 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35087 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42922 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18306 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36473 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16019 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37514 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46409 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38549 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37844 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41753 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17164 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14145 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40354 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18783 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9506 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18845 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18428 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -698,7 +698,6 @@ void RenderLayerModelObject::repaintOrRelayoutAfterSVGTransformChange()
 
     // Instead of performing a full-fledged layout (issuing repaints), just recompute the layer transform, and repaint.
     // In LBSE transformations do not affect the layout (except for text, where it still does!) -- SVG follows closely the CSS/HTML route, to avoid costly layouts.
-    updateLayerTransform();
```
Good catch.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1332,6 +1332,15 @@ void RenderLayer::updateTransform()
     if (hasTransform) {
         m_transform->makeIdentity();
         updateTransformFromStyle(*m_transform, renderer().style(), RenderStyle::allTransformOperations());
+        if (!hadTransform && !m_transform->isIdentity()) {
```
Hmm, this is unncessary for non-SVG renderers at least.
Also extra work if CSS transforms are used for SVG renderers. I think the new code updating the stacking context state needs to be performed _only_ when we change the transform attr/prop of a SVG renderer + repaint, instead of a full-fledged layout, no?

rwlbuis replied:
You are right, the location I picked was not very SVG specific, I just did not like adding a method on RenderLayer, but I guess we can't avoid that. New iteration should be better, PTAL :)

rwlbuis committed:
[LBSE] Fix svg/repaint/image-with-clip-path.svg
https://bugs.webkit.org/show_bug.cgi?id=270187

Reviewed by Nikolas Zimmermann.

On a transform change from identity matrix to non-identity matrix
the render layer needs to become a stacking context without relying on
a style update in the case of LBSE. Add logic for this in repaintOrRelayoutAfterSVGTransformChange.

Also add a correction to only call updateLayerTransform once in
repaintOrRelayoutAfterSVGTransformChange.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateTransform):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::repaintOrRelayoutAfterSVGTransformChange):

Canonical link: https://commits.webkit.org/275683@main
